proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+        public GetRequestedData<string> GetRequestedDataUsingParentId { get; set; }
+        public GetRequestedData<ActivityContext> GetRequestedDataUsingContext { get; set; }
+        public void Dispose() { }
+    }    
+}
+```
+
+### ActivitySource
+
+```C#
+namespace System.Diagnostics
+{
+    /// ActivitySource is the type that help create and start Activity and allow listening to the Activity events too.
+    public sealed class ActivitySource : IDisposable
+    {
+        /// Construct an ActivitySource object with the input name and component version
+        public ActivitySource(string name, string version = "");
+
+        /// Returns the ActivitySource name.
+        public string Name { get; }
+
+        /// Returns the ActivitySource version.
+        public string Version { get; }
+
+        /// Check if there is any listeners for this ActivitySource.
+        /// This property can be helpful to tell if there is no listener, then no need to create Activity object
+        /// and avoid creating the objects needed to create Activity (e.g. ActivityContext)
+        /// Example of that is http scenario which can avoid reading the context data from the wire.
+        public bool HasListeners();
+
+        /// Creates a new Activity object if there is any listener to the Activity, returns null otherwise.
+        public Activity? StartActivity(string name, ActivityKind kind = ActivityKind.Internal);
+
+        /// Creates a new <see cref="Activity"/> object if there is any listener to the Activity events, returns null otherwise.
+        public Activity? StartActivity(
+                            string name,
+                            ActivityKind kind,
+                            ActivityContext parentContext, 
+                            IEnumerable<KeyValuePair<string, string?>>? tags = null,
+                            IEnumerable<ActivityLink>? links = null,
+                            DateTimeOffset startTime = default);
+
+        /// Creates a new <see cref="Activity"/> object if there is any listener to the Activity events, returns null otherwise.
+        public Activity? StartActivity(
+                            string name,
+                            ActivityKind kind,
+                            string parentId,
+                            IEnumerable<KeyValuePair<string, string?>>? tags = null,
+                            IEnumerable<ActivityLink>? links = null,
+                            DateTimeOffset startTime = default);
+
+        /// Dispose the ActivitySource object and remove the current instance from the global list. empty the listeners list too.
+        public void Dispose();
+
+        /// Add a listener to the Activity starting and stopping events.
+        public static void AddActivityListener(ActivityListener listener) { }
+    }
+```
+
+### Activity
+
+```C#
+namespace System.Diagnostics
+{
+    public partial class Activity : IDisposable // added IDisposable
```

Comment:
Was Activity partial class before? 
can you explain the reasoning behind it?

Reply:
Partial is just C# notion to allow writing the class code in multiple source files (and not in single file). The reason we use partial is we have some code specific to netfx, netstandard, and netcoreapp. instead of filling the code with a lot of #ifdef, we just factor the code to its own source files and marking the class as partial so the compiler understand it is just one class at the end.